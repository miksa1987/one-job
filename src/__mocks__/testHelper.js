export const saveTodoMock = jest.fn();