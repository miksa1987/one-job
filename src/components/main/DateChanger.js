import React from 'react';
import styled from 'styled-components';
import moment from 'moment';
import MarginlessButton from '../common/MarginlessButton';
import MarginlessInput from '../common/MarginlessInput';
import TodoStore from '../../store';
import useField from '../../hooks/useField';
import { observer } from 'mobx-react-lite';

const DateChangerBase = styled.div`
  display: grid;
  grid-template: 2rem / 3rem 3rem 3rem 6rem 3rem 6rem;
`;

const DateChanger = observer(() => {
  const [day, setDay] = useField('text');
  const [month, setMonth] = useField('text');
  const [year, setYear] = useField('text');

  const store = React.useContext(TodoStore);

  React.useEffect(() => {
    const dateNow = moment().format('YYYY-MM-DD');
    setDate(dateNow);
  }, []);

  const previousDay = () => {
    const previousDayDate = moment(store.date, 'YYYY-MM-DD').add(-1, 'day').format('YYYY-MM-DD');
    setDate(previousDayDate);
  }

  const nextDay = () => {
    const nextDayDate = moment(store.date, 'YYYY-MM-DD').add(1, 'day').format('YYYY-MM-DD');
    setDate(nextDayDate);
  }

  const currentDay = () => {
    const currentDayDate = moment().format('YYYY-MM-DD');
    setDate(currentDayDate);
  }

  const setDate = (date) => {
    store.setCurrentDate(date);
    const splittedDate = date.split('-');
    
    setYear(splittedDate[0]);
    setMonth(splittedDate[1]);
    setDay(splittedDate[2]);
  }

  return (
    <DateChangerBase>
        <MarginlessButton onClick={previousDay}>{'<'}</MarginlessButton>
        <MarginlessInput {...day} />
        <MarginlessInput {...month} />
        <MarginlessInput {...year} />
        <MarginlessButton onClick={nextDay}>{'>'}</MarginlessButton>
        <MarginlessButton onClick={currentDay}>Current date</MarginlessButton>
    </DateChangerBase>
  );
});

export default DateChanger;