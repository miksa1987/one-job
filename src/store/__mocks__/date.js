export default class DateHandlerMock {
  setDate = jest.fn();
  setNextDayDate = jest.fn();
  setPreviousDayDate = jest.fn();
  setCurrentDayDate = jest.fn();
}