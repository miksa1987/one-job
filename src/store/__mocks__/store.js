import { decorate, computed, observable } from 'mobx';
import { createContext } from 'react';
import { 
  saveTodoMock, 
  setCurrentDateMock,
  loginAndSetUserMock,
  createAndSetUserMock,
  setNotificationMock 
} from './testHelper';

class TodoStore {
  currentTodo = {
    task: 'test',
    reflect: 'test',
    date: '2019-11-13',
    time: [ 12, 0 ],
    key: '1'
  };
  currentDate = '2019-11-13';

  setCurrentDate = (date) => {
    setCurrentDateMock();
    this.currentDate = date;
  }

  saveTodo = (newTodo, key) => {
    saveTodoMock();
    this.currentTodo = { ...newTodo, key: '1' };
  }

  setNextDayDate = () => this.currentDate = '2019-11-14';
  setPreviousDayDate = () => this.currentDate = '2019-11-12';
  setCurrentDayDate = () => this.currentDate = '2019-11-13';
  
  seeIfTodoTimeIsPassed = () => {}
  setNotification = () => setNotificationMock();

  loginAndSetUser = () => loginAndSetUserMock();
  createAndSetUser = () => createAndSetUserMock();

  get user() {
    return {};
  }
  
  get todo() {
    return this.currentTodo;
  }

  get date() {
    return this.currentDate;
  }

  get allTodos() {
    return [ this.currentTodo, this.currentTodo, this.currentTodos ];
  }

  get todoTimePassed() {
    return true;
  }
}

decorate(TodoStore, {
  currentUser: observable,
  currentTodo: observable,
  currentDate: observable,
  todos: observable,
  user: computed,
  todo: computed,
  date: computed,
  allTodos: computed
});

export default createContext(new TodoStore());