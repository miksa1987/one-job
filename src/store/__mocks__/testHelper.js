export const saveTodoMock = jest.fn();
export const setCurrentDateMock = jest.fn();
export const createAndSetUserMock = jest.fn();
export const loginAndSetUserMock = jest.fn();
export const setNotificationMock = jest.fn();