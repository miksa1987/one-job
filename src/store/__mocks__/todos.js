export default class TodosHandlerMock {
  getAndSetTodos = jest.fn();
  setCurrentTodo = jest.fn();
  checkForTodo = jest.fn();
  saveTodo = jest.fn();
  seeIfTodoTimeIsPassed = jest.fn();
}