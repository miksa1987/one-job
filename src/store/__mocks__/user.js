export default class UserHandlerMock {
  createAndSetUser = jest.fn();
  loginAndSetUser = jest.fn();
  logoutUser = jest.fn();
  setUser = jest.fn();
}