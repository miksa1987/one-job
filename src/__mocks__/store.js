import { decorate, computed, observable } from 'mobx';
import { createContext } from 'react';
import { saveTodoMock } from './testHelper';

class TodoStore {
  currentTodo = {
    task: 'test',
    reflect: 'test',
    date: '2019-11-13',
    time: [ 12, 0 ],
    key: '1'
  };

  saveTodo = (newTodo, key) => {
    saveTodoMock();
    this.currentTodo = { ...newTodo, key };
  }

  get user() {
    return {};
  }
  
  get todo() {
    return this.currentTodo;
  }

  get date() {
    return '2019-11-13';
  }

  get allTodos() {
    return [ this.currentTodo, this.currentTodo, this.currentTodos ];
  }
}

decorate(TodoStore, {
  currentUser: observable,
  currentTodo: observable,
  currentDate: observable,
  todos: observable,
  user: computed,
  todo: computed,
  date: computed,
  allTodos: computed
});

export default createContext(new TodoStore());