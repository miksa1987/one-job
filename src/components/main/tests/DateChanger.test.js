import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DateChanger from '../DateChanger';

import { setCurrentDateMock } from '../../../store/__mocks__/testHelper';
jest.mock('../../../store/store');

let date = [];

const renderAndGetComponentAndDate = () => {
  const component = render(<DateChanger />);
  const day = component.container.querySelector('#date-day');
  const month = component.container.querySelector('#date-month');
  const year = component.container.querySelector('#date-year');

  return [ component, day, month, year ];
}

describe('DateChanger', () => {
  afterEach(cleanup);

  test('Date is rendered', () => {
    const [ component, day, month, year ] = renderAndGetComponentAndDate();

    expect(day).toBeDefined();
    expect(month).toBeDefined();
    expect(year).toBeDefined();

    date = [ 
      Number(day.value), 
      Number(month.value), 
      Number(year.value)
    ];
  });

  test('Date can be changed forward', () => {
    const [ component, day, month, year ] = renderAndGetComponentAndDate();
    
    const nextDay = component.container.querySelector('#next-day');
    fireEvent.click(nextDay);

    expect(Number(day.value)).toBe(Number(date[0]) + 1);
  });

  test('Date can be changed backward', () => {
    const [ component, day, month, year ] = renderAndGetComponentAndDate();
    
    const previousDay = component.container.querySelector('#previous-day');
    fireEvent.click(previousDay);

    expect(Number(day.value)).toBe(Number(date[0]) - 1);
  });

  test('Date can be changed to current', () => {
    const [ component, day, month, year ] = renderAndGetComponentAndDate();

    const previousDay = component.container.querySelector('#previous-day');
    fireEvent.click(previousDay);
    
    const currentDate = component.container.querySelector('#current-date');
    fireEvent.click(currentDate);

    expect(Number(day.value)).toBe(Number(date[0]));
  });
});