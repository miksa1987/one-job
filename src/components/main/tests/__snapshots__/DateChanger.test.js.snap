// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateChanger Date is rendered 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-htpNat eoraiM"
      >
        <button
          class="sc-bdVaJa grEZJp"
          id="previous-day"
        >
          &lt;
        </button>
        <input
          class="sc-bwzfXH hEtCYp"
          id="date-day"
          type="text"
          value="13"
        />
        <input
          class="sc-bwzfXH hEtCYp"
          id="date-month"
          type="text"
          value="11"
        />
        <input
          class="sc-bwzfXH hEtCYp"
          id="date-year"
          type="text"
          value="2019"
        />
        <button
          class="sc-bdVaJa grEZJp"
          id="next-day"
        >
          &gt;
        </button>
        <button
          class="sc-bdVaJa grEZJp"
          id="current-date"
        >
          Current date
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-htpNat eoraiM"
    >
      <button
        class="sc-bdVaJa grEZJp"
        id="previous-day"
      >
        &lt;
      </button>
      <input
        class="sc-bwzfXH hEtCYp"
        id="date-day"
        type="text"
        value="13"
      />
      <input
        class="sc-bwzfXH hEtCYp"
        id="date-month"
        type="text"
        value="11"
      />
      <input
        class="sc-bwzfXH hEtCYp"
        id="date-year"
        type="text"
        value="2019"
      />
      <button
        class="sc-bdVaJa grEZJp"
        id="next-day"
      >
        &gt;
      </button>
      <button
        class="sc-bdVaJa grEZJp"
        id="current-date"
      >
        Current date
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
